docs/docs/vm-fundamentals/builtins.md

```diff
@@ -4,4 +4,229 @@ sidebar_position: 2
+    ```rust
+    // Function to add two numbers and write the sum to the output
+    func add(a felt, b felt) -> (felt) {
+      let sum = a + b
+      // Use the output builtin to write the sum to the program's output
+      output(sum)
+      return sum
+    }
+    ```
+    
+- **pedersen**
+    
+    The pedersen builtin, accessed with a pointer to type HashBuiltin, is used for pedersen hashing computations.
+    
+    ```rust
+    // Function to perform Pedersen hash on two numbers
+    // This function takes two felt values (a and b) as input.
+    func pedersen_hash(a felt, b felt) -> (HashBuiltin) {
+    	// Call the `pedersen` builtin to perform the Pedersen hash computation on the two inputs (a and b).
+      // The `pedersen` builtin is likely accessed with a pointer to a type named `HashBuiltin` 
+      // (specific names might vary depending on Cairo version).
+      
+      let hash = pedersen(a, b)
+      // The result of the Pedersen hash is stored in a variable named `hash`. 
+      // The type of `hash` is likely `HashBuiltin`.
+      
+      // Return the calculated hash value.
+      return hash
+    }
+    ```
+    
+- **range_check**
+    
+    Unlike other builtins, the range_check is accessed using a type felt, rather than a pointer. This builtin is mostly used for integer comparisons, and facilitates check to confirm that a field element is within a range `[0, 2^128)`
+    
+    ```rust
+    // Function to check if a number is within the valid range
+    func is_valid_range(x felt) -> (felt) {
+      // Use the range_check builtin to check the range
+      if !range_check(x) {
+        // Handle the case where the value is out of range (optional)
+      }
+      return x
+    }
+    ```
+    
+- **ecdsa**
+    
+    The ecdsa builtin, accessed with a pointer to type SignatureBuiltin, is used for verifying ECDSA signatures on a message using  public key.
+    
+    ```rust
+    // Function to verify an ECDSA signature (complex example)
+    func verify_signature(message felt, signature SignatureBuiltin, public_key felt) -> (felt) {
+      // This example requires additional logic to unpack the signature and public key
+      // and call the ecdsa builtin for verification. Refer to Cairo documentation for details.
+      let is_valid = ecdsa(message, signature, public_key)
+      // Handle the result of the verification (valid or invalid)
+      return is_valid
+    }
+    ```
+    
+- **bitwise**
+    
+    The bitwise builtin, accessed with a pointer to type BitwiseBuiltin, is used for carrying out bitwise operations on felts.
+    
+    ```rust
+    // Function to perform a bitwise AND operation on two numbers
+    func bitwise_and(a felt, b felt) -> (felt) {
+    //it checks if each bit position in a and b is a 1 at the same time  
+      let result = bitwise(BitwiseBuiltin.AND, a, b)
+      return result
+    }
+    ```
+    
+
+### How bulitins perform complex operations at the cost of extra constraints when proving
```

Comment:
Builtins come with a cost, but what you describe in this section is completely unrelated as far as my knowledge goes. I personally don't find the knowledge in this chapter beneficial for understanding builtins.